src/constraint_system.rs

```diff
@@ -156,6 +156,44 @@ pub trait ConstraintSystem<Scalar: PrimeField>: Sized + Send {
             "ConstraintSystem::extend must be implemented for types implementing ConstraintSystem"
         );
     }
+
+    /// ConstraintSystems that are witness generators need not assemble the actual constraints. Rather, they exist only
+    /// to efficiently create a witness.
+    fn is_witness_generator(&mut self) -> bool {
+        false
+    }
```

Comment:
Why does this take `&mut self` if this is a flag?

Reply:
For some reason, using `&self` everywhere leads to a compilation error. It's unclear to me what's going on.
```
error[E0596]: cannot borrow `*self` as mutable, as it is behind a `&` reference                                                                                                            |    fn is_witness_generator(&self) -> bool {                                                                                                                                                
   --> src/constraint_system.rs:265:9                                                                                                                                                      |        false                                                                                                                                                                               
    |                                                                                                                                                                                      |    }                                                                                                                                                                                       
162 |     fn is_witness_generator(&self) -> bool {                                                                                                                                         |                                                                                                                                                                                            
    |                             ----- help: consider changing that to be a mutable reference: `&mut self`                                                                                |    fn extend_inputs(&mut self, _new_inputs: &[Scalar]) {                                                                                                                                   
...                                                                                                                                                                                        |        assert!(!self.is_witness_generator());                                                                                                                                              
265 |         self.get_root().is_witness_generator()                                                                                                                                       |        unimplemented!()                                                                                                                                                                    
    |         ^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable                                                                            |    }
```

Reply:
Fix this and it will go away: https://github.com/filecoin-project/bellperson/pull/309#discussion_r1214400489

Reply:
Aha!

Reply:
However, my joy is short-lived. Making those changes does indeed fix the type problems here. But it causes downstream tests to fail.

```
--- STDERR:              neptune circuit2_witness::test::test_poseidon_hash_witness ---
thread 'circuit2_witness::test::test_poseidon_hash_witness' panicked at 'assertion failed: cs.is_witness_generator()', src/circuit2_witness.rs:32:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This makes me believe the change of implementation is *not* actually equivalent, even though it looks like it should be. This is the kind of complicated, picky, entangled … fun I've been griping about.

I'll be happy if it turns out I'm missing something obvious, though.

Reply:
Maybe I need to properly propagate this through to neptune. That failure is worrisome, though.

Reply:
Alas, even after applying your suggested changes to neptune, the error persists.

Reply:
What is going on, specifically, in the failure of `test_poseidon_hash_witness`, is the following:
- the first assertion (`assert!(cs.is_witness_generator())`) of [`poseidon_hash_allocated_witness`](https://github.com/lurk-lab/neptune/blob/47a96a7aef7e16caddd188178eabd2d8ea65c403/src/circuit2_witness.rs#L22-L26) fails,
- this function unavoidably[^1] is called on a `&mut wcs` where `wcs` is a `WitnessConstraintSystem` [here](https://github.com/lurk-lab/neptune/blob/47a96a7aef7e16caddd188178eabd2d8ea65c403/src/circuit2_witness.rs#L439-L440). You can of course check, *before* the call to `poseidon_hash_allocated_witness` that triggers the panic, that `wcs.is_witness_generator()` returns true,
- so the instance of `ConstraintSystem::is_witness_generator` cannot be that of the `WitnessConstraintSystem` of `wcs`, but instead it is this one, labeled as a convenience:
https://github.com/filecoin-project/bellperson/blob/37ac590ea3eab8ce76699f95e191ceaece928519/src/constraint_system.rs#L233-L288
- the reason why this returns `false` on `is_witness_generator`, in the case of our test, is that this implementation omits the `is_witness_generator` field, and therefore defers to the underlying default implementation situated in the trait definition,
- note that this also omits `is_extensible`, which is liable to exhibiting the same difference in behaviour for the flag depending on whether it is called on a value / reference, or on an `&mut`,
- the reason this would pass when the flag had the signature `is_witness_generator(&mut self)` is because this signature invites the type system prioritize a search for a method part of an instance of `ConstraintSystem` on `self`, before looking up one on (&mut self),[^2]
- While backwards-compatibility requires us to keep the `ConstraintSystem` for `&mut T`, I think its behavior is a clear bug w.r.t at least flags, and possibly for any other member where it defers to the default fields. Such a convenience implementation should not change *any* aspect of the instance of `ConstraintSystem` it is a proxy for [^3]
- So I would advise fixing this issue by making sure that convenience instance implements at least all flags (`is_extensible`, `is_witness_generator`) by delegation, 
- I would also recommend implementing the other missing members (`namespace`, `extend`, `extend_inputs`, `extend_aux`, `allocate_empty`, `allocate_empty_inputs`, `allocate_empty_aux`) and bumping a major version to indicate a change in behavior that may break backwards compatibility.

[^1]: See the signature of `poseidon_hash_witness_into_cs`
[^2]: if you must, look up 'autoref' for the priorities of method lookup, or explore the [compiler](https://github.com/rust-lang/rust/blob/master/compiler/rustc_hir_typeck/src/method/probe.rs)
[^3]: Other instances of `ConstraintSystem` that are derived from an instance of a `ConstraintSystem` can and do make other choices (e.g. `NameSpace`). But I think that may be OK, as long as we're aware of their behavior, since they don't claim to be a simple convenience that's easy to inadvertently call - as this bug shows.